Sources/Fuzzilli/Corpus/MarkovCorpus.swift

```diff
@@ -0,0 +1,196 @@
+            assert(programExecutionQueue.count > 0)
+            var prog : Program
+            if let tempProg = currentProg, remainingEnergy > 0 {
+                prog = tempProg
+                remainingEnergy -= 1
+            } else {
+                prog = programExecutionQueue.popLast()!
+                currentProg = prog
+                remainingEnergy = energyBase()
+            }
+            return prog
+        } else {
+            return allIncludedPrograms.randomElement()!
+        }
+    }
+
+    private func regenProgramList() {
+        assert(programExecutionQueue.count == 0)
+        let covEdgesBuffPtr = covEvaluator.getEdgeCountPtr()!
+        // Per https://developer.apple.com/documentation/swift/unsafebufferpointer, making an array from an UnsafeBufferPtr copies the memory
+        var covEdgesArr = Array(covEdgesBuffPtr)
+        covEdgesArr.sort()
+
+        // Find the edge with the smallest count
+        var startIndex = -1
+        for (i, val) in covEdgesArr.enumerated() {
+            if startIndex == -1 && val != 0 {
+                startIndex = i
+                break
+            }
+        }
+        assert(startIndex != -1)
+        
+        // Find the nth interesting edge's count
+        let desiredEdgeCount = max(size / 8, 30)
+        let endIndex = min(startIndex + desiredEdgeCount, covEdgesArr.count - 1)
+        let maxEdgeCountToFind = covEdgesArr[endIndex]
+        let amountToAge = energyBase() * numConsecutiveMutations // Likely overaggressive
+
+        // Find the n edges with counts <= maxEdgeCountToFind. Age them appropriately,
+        // to ensure non-deterministic samples are not continuously selected
+        for (i, val) in covEdgesBuffPtr.enumerated() {
+            // Applies dropout on otherwise valid samples, to ensure variety between instances
+            // This will likely select some samples multiple times, which is acceptable as
+            // it is proportional to how many infrquently hit edges the sample has
+            if val != 0 && val <= maxEdgeCountToFind && probability(0.25){ 
+                if let prog = edgeMap[UInt64(i)] {
+                    programExecutionQueue.append(prog)
+                    covEdgesBuffPtr[i] += amountToAge
+                } else {
+                    logger.warning("Failed to find edge in map")
+                }
+            }
+        }
+        assert(programExecutionQueue.count > 0)
+        logger.info("Markov Corpus selected \(programExecutionQueue.count) new programs")
+    }
+
+    public func exportState() throws -> Data {
+        let res = try encodeProtobufCorpus(allIncludedPrograms)
+        logger.info("Successfully serialized \(allIncludedPrograms.count) programs")
+        return res
+    }
+    
+    public func importState(_ buffer: Data) throws {
+        let newPrograms = try decodeProtobufCorpus(buffer)        
+        for prog in newPrograms { 
+            add(prog, ProgramAspects(outcome: .succeeded))
```

Comment:
So when synchronising workers now, this path is now taken, and workers won't have the `edgeMap` populated after a sync, correct?
I guess there are three possible solutions:
1. add the periodic "reimport" feature that you've mentioned, then run a reimport at the end of `importState`
2. Add a `evaluator.reset()` API, call this from here, then do a "slow" import where every program is executed and evaluated again
3. Also transfer the aspects. One option for this would be to attach the aspects to every `Program` instance, then serialize them to a protobuf. This option would also allow removing the `aspects` parameter of the `add` method again.
WDYT?

Reply:
Yes, this change definitely causes an issue with the edge map.

I think it can best be solved in a similar fashion to the non-determinism discussion [here](https://github.com/googleprojectzero/fuzzilli/pull/171#discussion_r544746061). Since we are already planning on doing a second execution for samples found by a particular instance, we could do two executions for samples imported by another instance. The downside here is that there would be a moderate startup time, but it would only be a small part of a serious fuzzing run.

A periodic re-import could provide additional value, in that it could remove samples that are subsumed by other ones, choosing based on size of the sample, execution speed, or some other metric.